google/cloud/storage/tests/curl_nosignal_integration_test.cc

```diff
@@ -0,0 +1,185 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/client.h"
+#include "google/cloud/storage/list_objects_reader.h"
+#include "google/cloud/storage/testing/storage_integration_test.h"
+#include "google/cloud/testing_util/assert_ok.h"
+#include "google/cloud/testing_util/init_google_mock.h"
+#include <gmock/gmock.h>
+#include <chrono>
+#include <future>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace {
+
+using ::google::cloud::storage::testing::TestPermanentFailure;
+using ::testing::ElementsAreArray;
+using ::testing::HasSubstr;
+
+/// Store the options captured from the command-line arguments.
+std::string FLAG_bucket_name;
+std::chrono::seconds FLAG_idle_duration;
+
+class CurlNoSignalIntegrationTest
+    : public google::cloud::storage::testing::StorageIntegrationTest {};
+
+Status UploadFiles(std::string const& media,
+                 std::vector<std::string> const& names) {
+  auto bucket_name = FLAG_bucket_name;
+  StatusOr<Client> client = Client::CreateDefaultClient();
+  if (!client) {
+    return client.status();
+  }
+
+  for (auto const& object_name : names) {
+    // Raise on error so the errors are reported to the thread that launched
+    // this function.
+    auto meta = client
+                              ->InsertObject(bucket_name, object_name, media,
+                                             IfGenerationMatch(0), Fields(""));
+    if (!meta) {
+      return meta.status();
+    }
+  }
+  return Status();
+}
+
+Status DownloadFiles(int iterations, std::vector<std::string> const& names) {
+  if (names.empty()) {
+    // Nothing to do, should not happen, but checking explicitly so the code is
+    // more readable.
+    return Status(StatusCode::kInvalidArgument, "empty object name list");
+  }
+
+  auto bucket_name = FLAG_bucket_name;
+  StatusOr<Client> client = Client::CreateDefaultClient();
+  if (!client) {
+    return client.status();
+  }
+
+  std::random_device rd;
+  std::mt19937_64 gen(rd());
```

Comment:
It is generally considered better practice to not recreate `gen` multiple times due to being an expensive object with a lot of state, so that may inadvertedly affect your results. I would make these global.

Reply:
Yup, they are big, but they are also not thread safe.  I am trading off expensive initialization once per thread vs. extra mutexes N times on each thread.  I am happy with that trade off.

The real downside is the draining of the entropy on `std::random_device`.  But that in itself is not as bad as you might think on modern processors (whatever the ones with the `RND` instruction are called).  Though it is a problem on VMs, which have terribly low entropy for more or less obvious reasons.

In any case, this is a test that will be executed manually, most likely on my workstation, so maybe: ¯\_(ツ)_/¯

Reply:
> I am trading off expensive initialization once per thread vs. extra mutexes N times on each thread.

Should `thread_local` not suffice here? See https://github.com/lichray/randint/blob/master/random.h#L49. But since this is a one-off test it's not that big of a problem.